### PR TITLE
Prepare for pip-packaging

### PR DESCRIPTION
**Pull request by bryant1410**
Fix allenai/allennlp#3612.

It adds a setup.py file for installation through pip.

I did the `configure` and `serve` subcommands, by subclassing `Subcommand`. For this, I'm supposing the allennlp main package will try to import allennlp-server, and if it's available, load the subcommands. Thoughts on this? We can later make that change in the main repository. I should also add a test of this behavior to this repository, I guess.

Some considerations:

* I used the file `setup.cfg` to define the metadata, as I believe it's clearer. I don't use `pyproject.toml` for it as I think it's a more drastic change (note I do use the latter for black).
* [`python setup.py test` is deprecated](https://setuptools.readthedocs.io/en/latest/history.html#v41-5-0), [so as `tests_require`](https://setuptools.readthedocs.io/en/latest/setuptools.html#new-and-changed-setup-keywords). That's why I didn't replicate this behavior from the main repo (nor the pytest runner thing). [PyPI recommends using tox](https://github.com/pypa/setuptools/issues/1684). However, I think it's not trivial to make each tox command a GitHub Action build step, so it'd be all one step, which doesn't help to debug/test. So I just kept it simple and didn't do anything about it.
* `zip_safe=False` from `setup` doesn't seem necessary here, as the `python setup.py` runs don't need to dynamically load any source file.

It's ready for review and discussion, but not to merge as it may need some more testing.

### CODE REVIEW

File: allennlp_server/commands/configure.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    $ allennlp configure --help
+    usage: allennlp configure [--help] [-h HOST] [-p PORT] [--include-package INCLUDE_PACKAGE]
+
+    Run a config explorer for allennlp.
+
+    optional arguments:
+      --help                show this help message and exit
+      -h HOST, --host HOST  interface to serve the wizard on (default = 127.0.0.1)
+      -p PORT, --port PORT  port to serve the wizard on (default = 8123)
+      --include-package INCLUDE_PACKAGE
```
Comment (bryant1410) [inline allennlp_server/commands/configure.py]: Note this option would be added from the main repository.

File: allennlp_server/commands/server_simple.py [hunk 2]
```diff
@@ -142,81 +162,79 @@ def _get_predictor(args: argparse.Namespace) -> Predictor:
...
-        field_names=field_names,
+        field_names=args.field_names,
         static_dir=args.static_dir,
         title=args.title,
     )
     CORS(app)
 
-    http_server = WSGIServer(("0.0.0.0", args.port), app)
-    print(f"Model loaded, serving demo on port {args.port}")
+    http_server = WSGIServer((args.host, args.port), app)
```
Comment (bryant1410) [inline allennlp_server/commands/server_simple.py]: I went from 0.0.0.0 to 127.0.0.1 as default for security reasons. If someone leaves it open with 0.0.0.0 when debugging, by accident it's exposing it potentially to the whole internet. I used an option for this so it can be changed back to 0.0.0.0 if needed.

On your side, if you folks are serving this repo, remember adding this option cause it breaks compatibility.

File: allennlp_server/config_explorer/config_explorer.py [hunk 3]
```diff
@@ -43,14 +41,10 @@ def to_dict(self):
...
 
-def make_app(include_packages: Sequence[str] = ()) -> Flask:
+def make_app() -> Flask:
     """
     Creates a Flask app that serves up a simple configuration wizard.
     """
-    # Load modules
-    for package_name in include_packages:
-        import_submodules(package_name)
-
```
Comment (bryant1410) [inline allennlp_server/config_explorer/config_explorer.py]: It'd be done by the main allennlp package.

File: allennlp_server/config_explorer/configuration.py [hunk 4]
```diff
@@ -36,6 +34,7 @@
 from allennlp.nn.regularizers import Regularizer
 from allennlp.training.optimizers import Optimizer as AllenNLPOptimizer
 from allennlp.training.trainer import Trainer
+from numpydoc.docscrape import NumpyDocString
```
Comment (bryant1410) [inline allennlp_server/config_explorer/configuration.py]: numpydoc should be removed down the road.

File: allennlp_server/tests/server_simple_test.py [hunk 5]
```diff
@@ -31,7 +28,8 @@ def setUp(self):
         super().setUp()
 
         print(os.getcwd())
-        archive = load_archive("tests/fixtures/bidaf/model.tar.gz")
+        importlib.import_module("allennlp_rc.models")
```
Comment (bryant1410) [inline allennlp_server/tests/server_simple_test.py]: This is used by `load_archive`, to find the "bidaf" model.

I think it's neater to load this like this, instead of the vanilla `import`, because the entity/object `BidirectionalAttentionFlow` is not used anywhere here, thus it's like the import it's not used. So I believe it's better to load it "internally", with `importlib`.

File: dev-requirements.txt [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+# Checks style, syntax, and other useful errors.
+flake8
+
+# Static type checking
+mypy==0.761
+
+pytest
+
+# Allows generation of coverage reports with pytest.
+pytest-cov
```
Comment (bryant1410) [inline dev-requirements.txt]: Same as in the main repo.

File: pyproject.toml [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+    | \.vscode
+    | \.venv
+    | \bdist\b
+    | \bdoc\b
+)
+'''
+
+[build-system]
+requires = ["setuptools", "wheel"]
+build-backend = "setuptools.build_meta"
```
Comment (bryant1410) [inline pyproject.toml]: Same as in the main repo.

File: setup.cfg [hunk 8]
```diff
@@ -0,0 +1,51 @@
+[metadata]
+name = allennlp-server
+version = 1.0.0-unreleased
```
Comment (bryant1410) [inline setup.cfg]: I'm supposing the version number here.

File: setup.cfg [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+  allennlp_reading_comprehension @ git+https://github.com/allenai/allennlp-reading-comprehension@c241f32
+
+  # Accessing files from S3 directly.
+  boto3
+
+  # REST interface for the models.
+  flask>=1.0.2
+  flask-cors>=3.0.7
+  gevent>=1.3.6
+python_requires = >=3.6.1
```
Comment (bryant1410) [inline setup.cfg]: Consistent with the main repo.

File: .github/workflows/main.yml [hunk 10]
```diff
@@ -15,17 +15,20 @@ jobs:
...
       run: |
         flake8 -v
         black -v --check .
     - name: Type check
       run: |
-        mypy config_explorer --ignore-missing-imports --no-strict-optional --no-site-packages
+        mypy allennlp_server --ignore-missing-imports --no-strict-optional --no-site-packages
+    - name: Check package setup
+      run: |
+        python setup.py check
```
Comment (bryant1410) [inline .github/workflows/main.yml]: This makes some sanity checks, such as required metadata or that the requirements are well specified.

File: allennlp_server/commands/configure.py [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+The ``configure`` subcommand launches a webapp that helps you generate an AllenNLP configuration file.
+
+    $ allennlp configure --help
+    usage: allennlp configure [--help] [-h HOST] [-p PORT] [--include-package INCLUDE_PACKAGE]
+
+    Run a config explorer for allennlp.
+
+    optional arguments:
+      --help                show this help message and exit
+      -h HOST, --host HOST  interface to serve the wizard on (default = 127.0.0.1)
```
Comment (bryant1410) [inline allennlp_server/commands/configure.py]: Using `-h` collides with `--help`, but I saw this is what `flask run` uses, so it could be expected by users. I don't have a strong opinion on this though.

File: Dockerfile [hunk 12]
```diff
@@ -1,20 +1,25 @@
...
+COPY dev-requirements.txt dev-requirements.txt
 RUN pip install -r requirements.txt
 
-COPY server_simple.py server_simple.py
-COPY configure.py configure.py
-COPY config_explorer/ config_explorer/
-COPY tests/ tests/
+COPY allennlp_server allennlp-server/
+
+RUN pip install -e .
```
Comment (DeNeutoy) [inline Dockerfile]: Looks like the entry point of this dockerfile needs updating

File: allennlp_server/commands/configure.py [hunk 13]
```diff
@@ -0,0 +1,54 @@
+"""
+The ``configure`` subcommand launches a webapp that helps you generate an AllenNLP configuration file.
+
+    $ allennlp configure --help
```
Comment (DeNeutoy) [inline allennlp_server/commands/configure.py]: Is this usage still correct? I'm not sure how allennlp will know to include this subcommand without something changing in the main repo? Maybe i'm missing the point...

File: allennlp_server/commands/server_simple.py (hunk 2 above)
Comment (DeNeutoy) [reply allennlp_server/commands/server_simple.py]: That's fine, thanks.

File: allennlp_server/config_explorer/config_explorer.py [hunk 14]
```diff
@@ -16,13 +16,11 @@
...
 
-from typing import Sequence
 import logging
 
-from flask import Flask, request, Response, jsonify, send_file
+from flask import Flask, jsonify, request, Response, send_file
 
-from config_explorer.configuration import configure, choices
-from allennlp.common.util import import_submodules
+from .configuration import choices, configure
```
Comment (DeNeutoy) [inline allennlp_server/config_explorer/config_explorer.py]: Preference for keeping the absolute imports to match our style

File: allennlp_server/commands/configure.py (hunk 13 above)
Comment (bryant1410) [reply allennlp_server/commands/configure.py]: Yeah, there are more details in the PR description. But the main idea is that the allennlp  main codebase checks for the presence of allennlp-server by importing it, and adds the subcommands. Does it make sense? Other ideas? Another idea that comes to my mind is a separate command.

Comment (matt-gardner) [reply allennlp_server/commands/configure.py]: Yes, I agree with this.  One option for doing this is to have something like a `.allennlp/plugins` file, where we import everything in that file before running our command.  This lets you add whatever package you want and removes the need for `--include-package`.  The pip install script could add a line to the plugin file.

File: Dockerfile [hunk 15]
```diff
@@ -1,22 +1,40 @@
 # This Dockerfile creates an environment suitable for downstream usage of AllenNLP.
-# It creates an environment that includes a pip installation of allennlp.
+# It creates an environment that includes a pip installation of allennlp-server.
 
-FROM python:3.6.8-stretch
+FROM python:3.6-stretch
```
Comment (bryant1410) [inline Dockerfile]: I think ideally this image should start with allennlp's. But I didn't find master pushed on Docker Hub. Thoughts?

File: allennlp_server/commands/configure.py (hunk 13 above)
Comment (bryant1410) [reply allennlp_server/commands/configure.py]: I searched a bit on this and found out that this is a thing: [Creating and discovering plugins](https://packaging.python.org/guides/creating-and-discovering-plugins/). The most convincing method to me is the 3rd one, [Using package metadata](https://packaging.python.org/guides/creating-and-discovering-plugins/#using-package-metadata), as the first one limits to packages starting with "allennlp" (maybe it's desired?) and the second one is complicated as it requires following certain directory hierarchy pattern and then adding packages.

So it'd be about declaring an entry point in the setup.py and then making allennlp import it, adding any available subcommand. Does it make sense? I'm gonna start working on it, and we can change it later if we decide.

**Comment by bryant1410** [comment]
Okay, now some code should be added in allennlp that:

1. Checks for installed plugins.
2. Imports them.
3. Load all the available subcommands, by looking at `Subcommand` subclasses.

Also, I think `Subcommand` should define a field with its own name (e.g., `name = "serve"`), to avoid allennlp knowing the name of every possible subcommand.

File: allennlp_server/commands/configure.py (hunk 13 above)
Comment (bryant1410) [reply allennlp_server/commands/configure.py]: I've analyzed it more. The problem I see with `.allennlp/plugins` is that any Python installation and virtual env in the system would read the same file, and for some of them, the plugins may not be installed.

For the solution I'm proposing, to remove the need for `--include-package`, the AllenNLP user has to include its code as a plugin, which involves installing its own package, which is cumbersome.

I think that using a standard namespace for it would be ideal. So, people can put some imports under a file in a certain folder to have their files discovered automatically. Python doesn't seem to support this without some packaging setup. We can import whatever is inside a certain folder, [like this old SO answer](https://stackoverflow.com/a/932072/1165181). I also found a library called [pluginlib](https://github.com/Rockhopper-Technologies/pluginlib), which seems to be somewhat used and maintained, which facilitates this. `Subcommand` can be the plugin interface, and then any subclasses are loaded. To also implement the `--include-package` behavior, other classes such as `Model` could also be plugins.

We can try just doing the first thing (defining a standard folder such as `allennlp.plugins` and loading whatever is there) and see how it goes.

Thoughts?

Comment (matt-gardner) [reply allennlp_server/commands/configure.py]: I think all we need to do is try to import every module in the plugins file. It's one package name per line, as a string. We just call exactly the same code that the current `--include-package` logic calls. If the import fails, due to a path error or a virtualenv issue as you brought up, we just give a big warning and move on. It's the job of the person who designed the package to make sure that everything that should be registered gets imported. And if our `Subcommands` aren't currently loaded from the registry, they should be.

This simple solution solves all of the issues, doesn't it?

Comment (matt-gardner) [reply allennlp_server/commands/configure.py]: I guess there's the question of how a line gets added to the plugins file on pip install, but I assume that's solvable. If a user wants to avoid using `--include-package` all the time, they can manually add a line to the plugins file.

Comment (bryant1410) [reply allennlp_server/commands/configure.py]: > It's the job of the person who designed the package to make sure that everything that should be registered gets imported. And if our Subcommands aren't currently loaded from the registry, they should be.

Agree.

> This simple solution solves all of the issues, doesn't it?

Yes, but one new issue I see: getting warnings in one virtualenv that uses allennlp for a plugin installed within another one. We could avoid showing warnings but silent errors are worse than noise when you want the feature to work.

I think that instead of trying to import from this global place, we could try to import all that is under `allennlp.plugins`, whatever is there. The core idea is the same as you say but with things local to the virtualenv instead of global to the system.

Comment (matt-gardner) [reply allennlp_server/commands/configure.py]: How does something get added to `allennlp.plugins`?  How does a user see what plugins are there, and add one manually?

Comment (bryant1410) [reply allennlp_server/commands/configure.py]: I'm gonna try and see if it works, but to add: add something to the folder (python package) `allennlp/plugins/`. To see what plugins are there: look for submodules of `allennlp.plugins`.

Comment (matt-gardner) [reply allennlp_server/commands/configure.py]: I would strongly prefer the other approach. I don't know where I would even go to find that folder in my environment, I'd have to dig around, so we can't give consistent instructions to users.  And those are automatically modified by pip in ways that I couldn't tell you, so I don't know if we would get clobbered on an update, or something.  *Much* better to just have a plain text file in a consistent location.

Comment (bryant1410) [reply allennlp_server/commands/configure.py]: Okay.

Just as a comment, I tested what I said but [it doesn't work with regular packages, only with namespace packages](https://docs.python.org/3/reference/import.html#packages), which is [one way Python recommends for plugin discoverability](https://packaging.python.org/guides/creating-and-discovering-plugins/#using-namespace-packages). But it seems users would need to [have a setup.py and use the attribute `packages`](https://packaging.python.org/guides/packaging-namespace-packages/#native-namespace-packages).

**Comment by DeNeutoy** [comment]
Sorry, I missed the discussion of this over the weekend.

A couple of things to add:

@bryant1410 - could you remove this commit https://github.com/allenai/allennlp-server/pull/3/commits/a85aa7a3f081f910a7112ed2f81b4e3249dcc9ef from this PR, so that we can merge it? The plugins is out of scope from just making this package pip installable (which is great!)


I am in favour of @matt-gardner's approach, with a couple of modifications:

The plugins file should not be global - in fact, I think it could be per working directory, like vscode have a `.vscode` folder where they put configuration.

It should support 3 things:

- packages, like this one
- individual paths to files/directories
- relative paths (maybe)


This way, we can have an allennlp command, `allennlp add <path>` which creates this file in the current working directory, or appends to it if it's there. This way is maybe a bit more verbose, but more tightly scoped for different projects. Thoughts?

**Comment by bryant1410** [comment]
@DeNeutoy I can remove it. But how can people use this repo? How do they use the commands?

I'm gonna think more about what you said about the local folder.

**Comment by matt-gardner** [comment]
I think the local folder could work; I don't have a strong opinion about that one way or the other.  Harder to just `pip install allennlp-server` and have it be available, though.  You would need to have another step in there between `pip install` and `allennlp configure`.

**Comment by DeNeutoy** [comment]
@bryant1410 - reasonably easily - they can just run the server as a script. It's ok to leave the commands there, just drop the plugin specific bit and then this looks good to me.

**Comment by DeNeutoy** [comment]
Actually, why do we need these to be sub-commands at all? I am very happy with these commands just being run like `python allennlp_server.configure --include-package ./my_project` . Feel a bit like we're bending over backward just to be able to run `allennlp configure`?

**Comment by matt-gardner** [comment]
On why do plugins at all, I think it's a general problem that could use a solution, especially as we have more sub-repos that break out specific functionality.  Adding a `.allennlp/plugins` file and a few lines of code to handle imports isn't really a lot of work, or a lot of complexity, and it gives tangible user benefits (removing the need for `--include-package`; having a simpler, unified command line API).  I agree that if the solution is incredibly complicated, it might not be worth the effort.  But my solution really isn't.

**Comment by DeNeutoy** [comment]
No, I am not talking about the plugins for adding packages and files to the registry, I think that is clearly useful. I am wondering about why we need to add things to the allennlp commands.

**Comment by matt-gardner** [comment]
For this particular case, because it's a first-party repo, and it simplifies the user experience to have commands from here be added to the typical place.  They also were available like that in the last major release, and this preserves compatibility.

For "bending over backwards" on this point, literally the only thing we need to do is change these lines to look in the registry: https://github.com/allenai/allennlp/blob/876d838b6eabce82225577a7938df3792012da68/allennlp/commands/__init__.py#L66-L80.  Doesn't seem like much work in order to get a unified (and extensible) command-line API?

**Comment by bryant1410** [comment]
@DeNeutoy I reversed that commit back. I guess it's ready to review but I'm not sure what's the actual command people can use now to use it. Also, there are references of `allennlp serve` and `allennlp configure`. Should these be changed back?

**Comment by bryant1410** [comment]
> LGTM, can you fix the lint (you might need to add an ignore rule for unused imports in **init** files). It's fine if this is a little bit in flux as we figure out the plugin thing.

done

File: allennlp_server/config_explorer/configuration.py [hunk 16]
```diff
@@ -401,7 +401,7 @@ def _render(item: ConfigItem, indent: str = "") -> str:
     return rendered_item
 
 
-BASE_CONFIG = Config(
+BASE_CONFIG: Config = Config(
```
Comment (bryant1410) [inline allennlp_server/config_explorer/configuration.py]: I wonder why are these necessary for mypy.
